### PR TITLE
Alerting: Support values in notification templates

### PR DESCRIPTION
**Pull request by grobinson-grafana**
**What this PR does / why we need it**:

We have received a lot of feedback regarding the `ValueString` in alert notifications. Perhaps one of the most frequent complaints about ValueString is that it is difficult to read because it contains a lot of information, and the information is shown as a JSON-like string. Users have often asked how it can be templated and the answer is that it can't.

Until now users have been able to add custom annotations to their alert rules which contains values via the $values variable added in previous versions of Grafana. However, these custom annotations must be added for each of the user's alert rule, instead of once in a template that all of their alerts can be notified via.

This commit adds then the much requested feature to support values in notification templates. Users can then create a single template that prints the annotations, labels and values of their alerts in a format of their choice!

Here is an example of the default template with an instant vector:

	**Firing**

	Value: A=1
	Labels:
	 - alertname = Test
	 - grafana_folder = General Alerting
	Annotations:
	Source: http://localhost:3000/alerting/grafana/wXZdwwVVz/view
	Silence: http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTest&matcher=grafana_folder%3DGeneral+Alerting

Here is an example of the default template with a reduce expression:

	**Firing**

	Value: B=1
	Labels:
	 - alertname = Test
	 - grafana_folder = General Alerting
	Annotations:
	Source: http://localhost:3000/alerting/grafana/wXZdwwVVz/view
	Silence: http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTest&matcher=grafana_folder%3DGeneral+Alerting

Here is an example of the default template with a reduce and math expression:

	**Firing**

	Value: B=1, C=1
	Labels:
	 - alertname = Test
	 - grafana_folder = General Alerting
	Annotations:
	Source: http://localhost:3000/alerting/grafana/wXZdwwVVz/view
	Silence: http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTest&matcher=grafana_folder%3DGeneral+Alerting

Last is an example of the default template with a non-reduced time series:

	**Firing**

	Value: [no value]
	Labels:
	 - alertname = DatasourceError
	 - grafana_folder = General Alerting
	 - rulename = Test
	Annotations:
	Source: http://localhost:3000/alerting/grafana/wXZdwwVVz/view
	Silence: http://localhost:3000/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DDatasourceError&matcher=grafana_folder%3DGeneral+Alerting&matcher=rulename%3DTest

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/state/cache.go [hunk 1]
```diff
@@ -38,6 +39,15 @@ func (c *cache) getOrCreate(ctx context.Context, alertRule *ngModels.AlertRule,
 
 	ruleLabels, annotations := c.expandRuleLabelsAndAnnotations(ctx, alertRule, result, extraLabels)
 
+	values := make(map[string]float64)
+	for _, v := range result.Values {
+		if v.Value != nil {
+			values[v.Var] = *v.Value
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/state/cache.go:45]: Just for my understanding, `v.Var` is the name of the stage, right? So i.e. "C" or "A" or whatever you call it.

Comment (grobinson-grafana) [reply pkg/services/ngalert/state/cache.go:45]: That's right! 😄

**Review by gillesdemey — APPROVED** [review]
LGTM! :shipit:

File: pkg/services/ngalert/notifier/channels/default_template.go [hunk 2]
```diff
@@ -16,8 +16,12 @@ const (
 var DefaultTemplateString = `
 {{ define "__subject" }}[{{ .Status | toUpper }}{{ if eq .Status "firing" }}:{{ .Alerts.Firing | len }}{{ if gt (.Alerts.Resolved | len) 0 }}, RESOLVED:{{ .Alerts.Resolved | len }}{{ end }}{{ end }}] {{ .GroupLabels.SortedPairs.Values | join " " }} {{ if gt (len .CommonLabels) (len .GroupLabels) }}({{ with .CommonLabels.Remove .GroupLabels.Names }}{{ .Values | join " " }}{{ end }}){{ end }}{{ end }}
 
+{{ define "__text_values_list" }}{{ $len := len .Values }}{{ if $len }}{{ $first := gt $len 1 }}{{ range $refID, $value := .Values -}}
```
Comment (gillesdemey) [inline pkg/services/ngalert/notifier/channels/default_template.go:19]: Does formatting of these lines affect the string output? If not I think we could add some whitespace and newlines to make this somewhat easier to read :)

Comment (grobinson-grafana) [reply pkg/services/ngalert/notifier/channels/default_template.go:19]: Not if we use `-` at the start or end of the braces, which removes whitespace. I did it at the end of each line, but wasn't sure if I should format it all like code considering most other templates are single line.

Comment (gillesdemey) [reply pkg/services/ngalert/notifier/channels/default_template.go:19]: Why not challenge the status quo – why are all the other ones on a single line? :D

**Review by yuri-tceretian — APPROVED** [review]
Can you please update the docs in UI 
image

here https://github.com/grafana/grafana/blob/4c4b7586464391d79f9c4eb18ab28ea12f73f79b/public/app/features/alerting/unified/components/receivers/TemplateData.ts#L114-L118

**Comment by jtheory** [comment]
Awesome, thank you; this'll be hugely helpful :trophy:

It looks to me like this doesn't risk breaking any existing alerting rule / templates -- the only difference it would make is that anyone who's currently dumping the entire ValueString into a notification will get a differently formatted result, right?

The main question I have is about updating docs & providing examples there, plus (ideally) how we should update the UI to help users 1) learn this is now possible  and 2) help them do it correctly
(UI updates aren't a blocker for this PR; I'm not sure if basic doc updates are a blocker or not)

**Pull request merged**